File: apps/medousa-home/src-tauri/src/daemon/vault.rs
use crate::daemon::types::{
    VaultBacklinksResponse, VaultChangesResponse, VaultFileContentResponse,
    VaultNoteContentResponse, VaultNotesListResponse, VaultRootsResponse, VaultSearchResponse,
    VaultTagsListResponse, VaultWriteResponse,
};
use medousa_types::{
    VaultAddRootRequest, VaultBacklinksQuery, VaultChangesQuery, VaultNotesQuery, VaultPutQuery,
    VaultSearchQuery, VaultSetActiveRootRequest, VaultTagsQuery, VaultWriteRequest,
};
use tauri::State;

use crate::embedded_daemon::EmbeddedDaemonState;

use super::sdk::{client, sdk_error};
use super::DaemonState;

fn encode_note_path(path: &str) -> String {
    path.split('/')
        .map(urlencoding::encode)
        .map(|segment| segment.into_owned())
        .collect::<Vec<_>>()
        .join("/")
}

#[tauri::command]
pub async fn vault_list_notes(
    state: State<'_, DaemonState>,
    _embedded_state: State<'_, EmbeddedDaemonState>,
    prefix: Option<String>,
    limit: Option<usize>,
    tags: Option<String>,
    tag_prefix: Option<String>,
    cursor: Option<String>,
    generation: Option<u64>,
) -> Result<VaultNotesListResponse, String> {
    let query = VaultNotesQuery {
        prefix: prefix.filter(|value| !value.trim().is_empty()),
        limit,
        tags: tags.filter(|value| !value.trim().is_empty()),
        tag_prefix: tag_prefix.filter(|value| !value.trim().is_empty()),
        cursor: cursor.filter(|value| !value.trim().is_empty()),
        generation,
    };
    #[cfg(target_os = "ios")]
    if let Some(client) = _embedded_state.client_if_active().await? {
        return client
            .list_vault_notes(query)
            .await
            .map_err(|error| error.to_string());
    }
    client(&state)?
        .vault()
        .list_notes(&query)
        .await
        .map_err(sdk_error)
}

#[tauri::command]
pub async fn vault_list_changes(
    state: State<'_, DaemonState>,
    _embedded_state: State<'_, EmbeddedDaemonState>,
    since_generation: Option<u64>,
    cursor: Option<String>,
    limit: Option<usize>,
) -> Result<VaultChangesResponse, String> {
    let query = VaultChangesQuery {
        since_generation,
        cursor: cursor.filter(|value| !value.trim().is_empty()),
        limit,
    };
    #[cfg(target_os = "ios")]
    if let Some(client) = _embedded_state.client_if_active().await? {
        return client
            .list_vault_changes(query)
            .await
            .map_err(|error| error.to_string());
    }
    client(&state)?
        .vault()
        .list_changes(&query)
        .await
        .map_err(sdk_error)
}

#[tauri::command]
pub async fn vault_list_tags(
    state: State<'_, DaemonState>,
    _embedded_state: State<'_, EmbeddedDaemonState>,
    prefix: Option<String>,
    limit: Option<usize>,
) -> Result<VaultTagsListResponse, String> {
    let query = VaultTagsQuery {
        prefix: prefix.filter(|value| !value.trim().is_empty()),
        limit,
    };
    #[cfg(target_os = "ios")]
    if let Some(client) = _embedded_state.client_if_active().await? {
        return client
            .list_vault_tags(query)
            .await
            .map_err(|error| error.to_string());
    }
    client(&state)?
        .vault()
        .list_tags(&query)
        .await
        .map_err(sdk_error)
}

#[tauri::command]
pub async fn vault_get_note(
    state: State<'_, DaemonState>,
    _embedded_state: State<'_, EmbeddedDaemonState>,
    path: String,
) -> Result<VaultNoteContentResponse, String> {
    #[cfg(target_os = "ios")]
    if let Some(client) = _embedded_state.client_if_active().await? {
        return client
            .get_vault_note(path.trim().to_string())
            .await
            .map_err(|error| error.to_string());
    }
    let encoded = encode_note_path(path.trim());
    client(&state)?
        .vault()
        .get_note(&encoded)
        .await
        .map_err(sdk_error)
}

#[tauri::command]
pub async fn vault_get_file(
    state: State<'_, DaemonState>,
    _embedded_state: State<'_, EmbeddedDaemonState>,
    path: String,
) -> Result<VaultFileContentResponse, String> {
    #[cfg(target_os = "ios")]
    if let Some(client) = _embedded_state.client_if_active().await? {
        return client
            .get_vault_file(path.trim().to_string())
            .await
            .map_err(|error| error.to_string());
    }
    let encoded = encode_note_path(path.trim());
    client(&state)?
        .vault()
        .get_file(&encoded)
        .await
        .map_err(sdk_error)
}

#[tauri::command]
pub async fn vault_save_note(
    state: State<'_, DaemonState>,
    _embedded_state: State<'_, EmbeddedDaemonState>,
    path: String,
    content: String,
    content_hash: Option<String>,
    session_id: Option<String>,
    auto_workshop_tags: Option<bool>,
) -> Result<VaultWriteResponse, String> {
    #[cfg(target_os = "ios")]
    if let Some(client) = _embedded_state.client_if_active().await? {
        let request = VaultWriteRequest {
            path: None,
            content,
            session_id: session_id.filter(|value| !value.trim().is_empty()),
            semantic_tags: None,
            auto_workshop_tags: auto_workshop_tags.unwrap_or(true),
        };
        return client
            .save_vault_note(
                path.trim().to_string(),
                request,
                content_hash.filter(|value| !value.trim().is_empty()),
            )
            .await
            .map_err(|error| error.to_string());
    }
    let encoded = encode_note_path(path.trim());
    let query = VaultPutQuery {
        session_id: session_id.filter(|value| !value.trim().is_empty()),
        auto_workshop_tags,
    };
    client(&state)?
        .vault()
        .update_note(
            &encoded,
            &content,
            &query,
            content_hash
                .as_deref()
                .filter(|value| !value.trim().is_empty()),
        )
        .await
        .map_err(sdk_error)
}

#[tauri::command]
pub async fn vault_create_note(
    state: State<'_, DaemonState>,
    _embedded_state: State<'_, EmbeddedDaemonState>,
    path: String,
    content: String,
    session_id: Option<String>,
    semantic_tags: Option<Vec<String>>,
    auto_workshop_tags: Option<bool>,
) -> Result<VaultWriteResponse, String> {
    let request = VaultWriteRequest {
        path: Some(path.trim().to_string()),
        content,
        session_id: session_id.filter(|value| !value.trim().is_empty()),
        semantic_tags: semantic_tags.filter(|tags| !tags.is_empty()),
        auto_workshop_tags: auto_workshop_tags.unwrap_or(true),
    };
    #[cfg(target_os = "ios")]
    if let Some(client) = _embedded_state.client_if_active().await? {
        return client
            .create_vault_note(request)
            .await
            .map_err(|error| error.to_string());
    }
    client(&state)?
        .vault()
        .create_note(&request)
        .await
        .map_err(sdk_error)
}

#[tauri::command]
pub async fn vault_delete_note(
    state: State<'_, DaemonState>,
    _embedded_state: State<'_, EmbeddedDaemonState>,
    path: String,
) -> Result<serde_json::Value, String> {
    #[cfg(target_os = "ios")]
    if let Some(client) = _embedded_state.client_if_active().await? {
        return client
            .delete_vault_note(path.trim().to_string())
            .await
            .map(|response| serde_json::to_value(response).unwrap_or_default())
            .map_err(|error| error.to_string());
    }
    let encoded = encode_note_path(path.trim());
    client(&state)?
        .vault()
        .delete_note(&encoded)
        .await
        .map(|response| serde_json::to_value(response).unwrap_or_default())
        .map_err(sdk_error)
}

#[tauri::command]
pub async fn vault_search(
    state: State<'_, DaemonState>,
    _embedded_state: State<'_, EmbeddedDaemonState>,
    query: String,
    limit: Option<usize>,
    tags: Option<String>,
) -> Result<VaultSearchResponse, String> {
    let trimmed = query.trim();
    let search = VaultSearchQuery {
        q: if trimmed.is_empty() {
            None
        } else {
            Some(trimmed.to_string())
        },
        limit: limit.or(Some(20)),
        tags: tags.filter(|value| !value.trim().is_empty()),
    };
    #[cfg(target_os = "ios")]
    if let Some(client) = _embedded_state.client_if_active().await? {
        return client
            .search_vault(search)
            .await
            .map_err(|error| error.to_string());
    }
    client(&state)?
        .vault()
        .search(&search)
        .await
        .map_err(sdk_error)
}

#[tauri::command]
pub async fn vault_backlinks(
    state: State<'_, DaemonState>,
    _embedded_state: State<'_, EmbeddedDaemonState>,
    path: String,
) -> Result<VaultBacklinksResponse, String> {
    #[cfg(target_os = "ios")]
    if let Some(client) = _embedded_state.client_if_active().await? {
        return client
            .vault_backlinks(path.trim().to_string())
            .await
            .map_err(|error| error.to_string());
    }
    let query = VaultBacklinksQuery {
        path: Some(path.trim().to_string()),
    };
    client(&state)?
        .vault()
        .backlinks(&query)
        .await
        .map_err(sdk_error)
}

#[tauri::command]
pub async fn vault_list_roots(
    state: State<'_, DaemonState>,
    _embedded_state: State<'_, EmbeddedDaemonState>,
) -> Result<VaultRootsResponse, String> {
    #[cfg(target_os = "ios")]
    if let Some(client) = _embedded_state.client_if_active().await? {
        return client.list_vault_roots().map_err(|error| error.to_string());
    }
    client(&state)?
        .vault()
        .list_roots()
        .await
        .map_err(sdk_error)
}

#[tauri::command]
pub async fn vault_set_active_root(
    state: State<'_, DaemonState>,
    _embedded_state: State<'_, EmbeddedDaemonState>,
    root_id: String,
) -> Result<VaultRootsResponse, String> {
    #[cfg(target_os = "ios")]
    if let Some(client) = _embedded_state.client_if_active().await? {
        return client
            .set_active_vault_root(root_id.trim())
            .map_err(|error| error.to_string());
    }
    let request = VaultSetActiveRootRequest {
        root_id: root_id.trim().to_string(),
    };
    client(&state)?
        .vault()
        .set_active_root(&request)
        .await
        .map_err(sdk_error)
}

#[tauri::command]
pub async fn vault_add_root(
    state: State<'_, DaemonState>,
    _embedded_state: State<'_, EmbeddedDaemonState>,
    label: String,
    path: String,
    id: Option<String>,
) -> Result<VaultRootsResponse, String> {
    #[cfg(target_os = "ios")]
    if let Some(client) = _embedded_state.client_if_active().await? {
        return client
            .add_vault_root(label.trim(), path.trim(), id.as_deref())
            .map_err(|error| error.to_string());
    }
    let request = VaultAddRootRequest {
        label: label.trim().to_string(),
        path: path.trim().to_string(),
        id: id.filter(|value| !value.trim().is_empty()),
    };
    client(&state)?
        .vault()
        .add_root(&request)
        .await
        .map_err(sdk_error)
}

#[tauri::command]
pub async fn vault_list_trash(
    state: State<'_, DaemonState>,
    _embedded_state: State<'_, EmbeddedDaemonState>,
    limit: Option<usize>,
) -> Result<serde_json::Value, String> {
    #[cfg(target_os = "ios")]
    if let Some(client) = _embedded_state.client_if_active().await? {
        return client
            .list_vault_trash(limit.unwrap_or(100))
            .await
            .map(|response| serde_json::to_value(response).unwrap_or_default())
            .map_err(|error| error.to_string());
    }
    client(&state)?
        .vault()
        .list_trash(limit)
        .await
        .map_err(sdk_error)
}

#[tauri::command]
pub async fn vault_restore_trash(
    state: State<'_, DaemonState>,
    _embedded_state: State<'_, EmbeddedDaemonState>,
    path: String,
) -> Result<serde_json::Value, String> {
    #[cfg(target_os = "ios")]
    if let Some(client) = _embedded_state.client_if_active().await? {
        return client
            .restore_vault_trash(path.trim().to_string())
            .await
            .map(|response| serde_json::to_value(response).unwrap_or_default())
            .map_err(|error| error.to_string());
    }
    client(&state)?
        .vault()
        .restore_trash(path.trim())
        .await
        .map_err(sdk_error)
}

File: apps/medousa-home/src/lib/daemon/vault.ts
import { invoke } from "@tauri-apps/api/core";
import type {
  VaultBacklinksResponse,
  VaultFileContentResponse,
  VaultNoteContentResponse,
  VaultNotesListResponse,
  VaultChangesResponse,
  VaultRootsResponse,
  VaultTagsListResponse,
  VaultSearchResponse,
  VaultWriteResponse,
} from "$lib/types/vault";

export async function listVaultNotes(options?: {
  prefix?: string;
  limit?: number;
  tags?: string[];
  tagPrefix?: string;
  cursor?: string;
  generation?: number;
}): Promise<VaultNotesListResponse> {
  const tags =
    options?.tags?.map((tag) => tag.trim()).filter(Boolean).join(",") || undefined;
  return invoke<VaultNotesListResponse>("vault_list_notes", {
    prefix: options?.prefix,
    limit: options?.limit,
    tags,
    tagPrefix: options?.tagPrefix,
    cursor: options?.cursor,
    generation: options?.generation,
  });
}

export async function listVaultChanges(options?: {
  sinceGeneration?: number;
  cursor?: string;
  limit?: number;
}): Promise<VaultChangesResponse> {
  return invoke<VaultChangesResponse>("vault_list_changes", {
    sinceGeneration: options?.sinceGeneration,
    cursor: options?.cursor,
    limit: options?.limit,
  });
}

export async function listVaultTags(options?: {
  prefix?: string;
  limit?: number;
}): Promise<VaultTagsListResponse> {
  return invoke<VaultTagsListResponse>("vault_list_tags", {
    prefix: options?.prefix,
    limit: options?.limit,
  });
}

export async function getVaultNote(
  path: string,
): Promise<VaultNoteContentResponse> {
  return invoke<VaultNoteContentResponse>("vault_get_note", { path });
}

export async function getVaultFile(
  path: string,
): Promise<VaultFileContentResponse> {
  return invoke<VaultFileContentResponse>("vault_get_file", { path });
}

export async function saveVaultNote(
  path: string,
  content: string,
  options?: {
    contentHash?: string;
    sessionId?: string;
    autoWorkshopTags?: boolean;
  },
): Promise<VaultWriteResponse> {
  return invoke<VaultWriteResponse>("vault_save_note", {
    path,
    content,
    contentHash: options?.contentHash,
    sessionId: options?.sessionId,
    autoWorkshopTags: options?.autoWorkshopTags,
  });
}

export async function createVaultNote(
  path: string,
  content: string,
  options?: {
    sessionId?: string;
    semanticTags?: string[];
    autoWorkshopTags?: boolean;
  },
): Promise<VaultWriteResponse> {
  return invoke<VaultWriteResponse>("vault_create_note", {
    path,
    content,
    sessionId: options?.sessionId,
    semanticTags: options?.semanticTags,
    autoWorkshopTags: options?.autoWorkshopTags,
  });
}

export async function deleteVaultNote(path: string): Promise<{ path: string; deleted: boolean }> {
  return invoke<{ path: string; deleted: boolean }>("vault_delete_note", { path });
}

export async function searchVaultNotes(
  query: string,
  limit?: number,
  tags?: string[],
): Promise<VaultSearchResponse> {
  const tagFilter =
    tags?.map((tag) => tag.trim()).filter(Boolean).join(",") || undefined;
  return invoke<VaultSearchResponse>("vault_search", {
    query,
    limit,
    tags: tagFilter,
  });
}

export async function getVaultBacklinks(
  path: string,
): Promise<VaultBacklinksResponse> {
  return invoke<VaultBacklinksResponse>("vault_backlinks", { path });
}

export async function listVaultRoots(): Promise<VaultRootsResponse> {
  return invoke<VaultRootsResponse>("vault_list_roots");
}

export async function setActiveVaultRoot(rootId: string): Promise<VaultRootsResponse> {
  return invoke<VaultRootsResponse>("vault_set_active_root", { rootId });
}

export async function addVaultRoot(
  label: string,
  path: string,
  id?: string,
): Promise<VaultRootsResponse> {
  return invoke<VaultRootsResponse>("vault_add_root", { label, path, id });
}

export type VaultTrashEntry = {
  path: string;
  trashedAt?: string | null;
};

export async function listVaultTrash(limit?: number): Promise<{ entries: VaultTrashEntry[] }> {
  return invoke("vault_list_trash", { limit });
}

export async function restoreVaultTrash(
  path: string,
): Promise<{ path: string; restored: boolean }> {
  return invoke("vault_restore_trash", { path });
}

export type VaultGitDetect = {
  available: boolean;
  path?: string | null;
  version?: string | null;
  enabled: boolean;
  platformHint: string;
};

export type VaultGitStatus = {
  enabled: boolean;
  available: boolean;
  isRepo: boolean;
  branch?: string | null;
  dirtyCount: number;
  vaultRoot: string;
  gitPath?: string | null;
};

export type VaultGitLogEntry = {
  id: string;
  shortId: string;
  message: string;
  author: string;
  committedAt: string;
};

export async function vaultGitDetect(): Promise<VaultGitDetect> {
  return invoke<VaultGitDetect>("vault_git_detect");
}

export async function vaultGitStatus(): Promise<VaultGitStatus> {
  return invoke<VaultGitStatus>("vault_git_status");
}

export async function vaultGitEnable(
  enabled: boolean,
  initIfNeeded = false,
): Promise<{ enabled: boolean; status: VaultGitStatus }> {
  return invoke("vault_git_enable", { enabled, initIfNeeded });
}

export async function vaultGitInit(): Promise<VaultGitStatus> {
  return invoke<VaultGitStatus>("vault_git_init");
}

export async function vaultGitInstall(): Promise<VaultGitDetect> {
  return invoke<VaultGitDetect>("vault_git_install");
}

export async function vaultGitLog(options?: {
  path?: string;
  limit?: number;
}): Promise<VaultGitLogEntry[]> {
  return invoke<VaultGitLogEntry[]>("vault_git_log", {
    path: options?.path,
    limit: options?.limit,
  });
}

export async function vaultGitCommit(
  message: string,
  paths?: string[],
): Promise<{ id: string; message: string }> {
  return invoke("vault_git_commit", { message, paths: paths ?? [] });
}

export async function vaultGitRestore(commit: string, path: string): Promise<void> {
  await invoke("vault_git_restore", { commit, path });
}

export async function vaultGitDiff(
  path: string,
  commit?: string,
): Promise<{ path: string; patch: string }> {
  return invoke("vault_git_diff", { path, commit });
}

export async function vaultGitWorktrees(): Promise<
  Array<{ path: string; head: string; branch?: string | null }>
> {
  return invoke("vault_git_worktrees");
}
